keps/sig-api-machinery/3488-cel-admission-control/README.md

```diff
@@ -0,0 +1,2636 @@
+| exclude                                  | Audit (level=None)           | phase 1, see below "Exclude matching"  |
+| apiVersion + kind                        |                              | TBD                                    |
+| NonResourceURLs                          | Audit/RBAC/P&F               | No                                     |
+| default / fallthrough matching           |                              | phase 2 see "Default matching" section |
+| user/userGroup                           | Audit                        | phase 2 see "Secondary Authz" section  |
+| user.Extra                               | (in WH AdmissionReview)      | phase 2 see "Secondary Authz" section  |
+| permissions (RBAC verb)                  | RBAC                         | phase 2 see "Secondary Authz" section  |
+
+WH = Admission webhooks, P&F = Priority and Fairness
+
+Match criteria must be declared in the `spec.match` field of policy
+configuration resources (see `ReplicaLimit` in the above example) and will be
+declared with API types in a format similar to admission webhooks, P&F, RBAC and
+Audit.
+
+Proposed matching mechanism:
+
+- Matching criteria is defined as an ordered list of match objects (similar to
+  [Audit
+  Policy](https://kubernetes.io/docs/reference/config-api/apiserver-audit.v1/#audit-k8s-io-v1-Policy))
+  - If a match object evaluates to true
+    - If the match object is an "exclude", skip validation
+    - Otherwise perform validation
+  - If all match objects in the list evaluate to false, skip validation
+- The criteria of each match object in the list is applied conjunctively (ANDed)
+
+<<[UNRESOLVED, maxsmythe, ?? ]>>
+Consider an alternative where instead of a list of match criteria, we have a
+single match criteria where it is possible to handle excludes. I.e. we sanely
+generalize how labelSelectors support excludes to all the other match criteria?
+<<[/UNRESOLVED]>>
+
+Examples:
+
+```yaml
+# Matches resources with {namespace a, label x=1} and {namespace b, label x=2}
+# but not {namespace a, label x=2} or {namespace b, label x=1}.
+match:
+- namespaces: ["a"]
+  labelSelector:
+  - key: x
+    operator: In
+    values: ["1"] 
+- namespaces: ["b"]
+  labelSelector:
+- key: x
+    operator: In
+    values: ["2"] 
+```
+
+```yaml
+# Matches {namespace a, label x=2}, {namespace b, label x=1}, {namespace b, label x=2}
+# but not {namespace a, label x=1}.
+match:
+- namespaces: ["a"]
+  labelSelector:
+  - key: x
+    operator: In
+    values: ["1"] 
+  effect: Exclude
+- namespaces: ["a", "b"]
+  labelSelector:
+  - key: x
+    operator: In
+    values: ["1", "2"] 
+```
+
+<<[UNRESOLVED jpbetz, maxsmythe, ?? ]>>
+Provide wildcard prefix support to namespaces, resource name or any other fields?
+Or better to defer this type of check for CEL expression evaluation?
+<<[/UNRESOLVED]>>
+
+<<[UNRESOLVED jpbetz, maxsmythe ]>>
+Offer both GVR and GVK matching?
```

Comment:
I don't have any compelling use cases and so I'll drop this. Matching scale can be handled via "*/scale" as a resource so it doesn't require kind matching, and the explicit wildcard has nice readability properties when expressed as a resource.

@maxsmythe I did notice that OPA gatekeeper matched on GVK instead of GVR, so this may impact future integration. Maintaining a mapping outside of k8s between GVK/GVR seems like trouble, so not sure what options to recommend but maybe we can discuss more separately.